### ISSUE TITLE
Error in monitoring module (Specio::Exception) after upgrade to 2.020

### ISSUE THREAD

**Issue by YorVeX** [webmin/webmin#1860]
Just did the update on my Debian 11 system and now I get a root e-mail with this content every minute (every time the monitoring script is running I guess):

```
Validation failed for type named Num declared in package Specio::Library::Builtins (/usr/share/perl5/Specio/Library/Builtins.pm) at line 138 in sub named (eval) with value undef

Trace begun at Specio::Exception->new line 57
Specio::Exception::throw('Specio::Exception', 'message', 'Validation failed for type named Num declared in package Specio::Library::Builtins (/usr/share/perl5/Specio/Library/Builtins.pm) at line 138 in sub named (eval) with value undef', 'type', 'Specio::Constraint::Simple=HASH(0x55666a3fd108)', 'value', undef) called at (eval 247) line 88
DateTime::_check_from_epoch_params('locale', 'en-US', 'epoch', undef, 'time_zone', 'Europe/Berlin') called at /usr/lib/x86_64-linux-gnu/perl5/5.32/DateTime.pm line 481
DateTime::from_epoch('DateTime', 'locale', 'en-US', 'epoch', undef, 'time_zone', 'Europe/Berlin') called at ../web-lib-funcs.pl line 2093
WebminCore::make_date(undef) called at change-monitor.pl line 22
main::get_change_status('HASH(0x556669ac63b0)', 'change', undef) called at status-lib.pl line 181
eval {...} at status-lib.pl line 143
main::service_status('HASH(0x556669ac63b0)') called at /usr/share/webmin/status/monitor.pl line 75
```

I still get an e-mail when the status of a monitoring item changes, but the status in the overview isn't updated, e.g. still shows a green checkmark after a service has gone down.

Ideas welcome. Is it a generic problem that more people have or something specific to my monitoring configuration? I hope it's not, I got 76 monitoring targets and I really don't want to go through all of them and find out which one it is. Although I guess the most likely candidates would be one of those where I am making use of Regex expressions. The error message seems to refer to a DateTime type but I don't remember having any date specific checks in there.

**Comment by YorVeX** [comment]
I wanted to avoid switching from a repo installation to a manually installed 2.013 version now and having to go back later. I tried to downgrade through apt, but apparently the older version is already not in the repository anymore? That was fast, why not leave at least the previous version there for a while in case something goes wrong with an update?

Then I tried to disable scheduled monitoring until this is fixed so that I get rid of the e-mail spam, but I can't enter the options for this probably because of the same problem, after clicking the button the circle animation just never finishes and it never enters options:
![image](https://user-images.githubusercontent.com/528974/225467027-8f839874-fb0e-443a-aa10-8f3248af4618.png)

I once more learned my lesson not to be so fast with installing updates, at least when they are not marked as urgent security updates 😢

**Comment by jcameron** [comment]
I think the root cause of this is calling `make_date` on an undefined value. I've checked in a fix for that ...

**Comment by iliajie** [comment]
Oh, you're right! And unfortunately this won't be enough because there might be numerous other places in code where it's called that way. Therefore, I made another [patch](https://github.com/webmin/webmin/commit/88eb228f770f1c239b80d2ac6bf5b6a54fad00a7) to replicate the old behaviour where the "beginning" of time is returned, i.e. `01/01/1970 12:00 AM` in case of `undef` value.

Also, perhaps the better solution would be is to return an empty string?

**Comment by jcameron** [comment]
Yeah, replicating the old behavior seems like the safe option..

**Comment by YorVeX** [comment]
Thanks, with 2.021 monitoring is working again now (no error messages from the script execution, getting monitoring e-mails and the status is updated properly in the overview).

I still have the problem that nothing happens (except the circle animation) when clicking the "Scheduled Monitoring" button, but this might be a separate issue. It has been a while since I used this so I also can't say whether it was introduced with the 2.020 update for me or earlier (or with a webmin update at all or something I did).

When I enable the full webmin debug log for that module and click that button again it just does lots of status reads and finishes after 250 ms with - STOP "runtime=0" - which doesn't fit the picture in the browser where it's seemingly very busy with a task that never completes.

Should I open a separate issue for this?
